### PR TITLE
Backlog 20632 fix issues on set new canonical when previous one is locked

### PR DESCRIPTION
**Pull request by jahia-carp**
https://jira.jahia.org/browse/BACKLOG-20632

### CODE REVIEW

File: core/src/javascript/components/AddVanityUrl.jsx [hunk 1]
```diff
@@ -188,6 +191,26 @@ class AddVanityUrl extends React.Component {
             return entry;
         });
 
+        const urls = getRowUrlsFromPath(this.state.rows, path);
+        const lang = mappings[0].language;
+
+        // Exit if there is a canonical lock for deletion for the current lang
+        if (mappings[0].defaultMapping === true && atLeastOneCanonicalLockedForLang(urls, lang)) {
```
Comment (jsinovassin) [inline core/src/javascript/components/AddVanityUrl.jsx:198]: Could it be replaced by 
`if (mappings[0].defaultMapping && atLeastOneCanonicalLockedForLang(urls, lang)) {`
?

File: core/src/javascript/components/Utils/Utils.jsx [hunk 2]
```diff
@@ -21,6 +21,10 @@ export const atLeastOneMarkedForDeletion = urls => {
     return urls.some(url => url.default.publicationInfo.publicationStatus === 'MARKED_FOR_DELETION' || url.default.mixinTypes.find(mixin => mixin.name === 'jmix:markedForDeletion'));
 };
 
+export const atLeastOneCanonicalLockedForLang = (urls, lang) => {
+    return urls.some(url => url.default.mixinTypes.find(mixin => mixin.name === 'jmix:markedForDeletion') && url.default.language === lang && url.default.default === true);
```
Comment (jsinovassin) [inline core/src/javascript/components/Utils/Utils.jsx:25]: Maybe you could remove `=== true ` from `url.default.default === true`

File: core/src/javascript/components/Utils/Utils.jsx [hunk 3]
```diff
@@ -21,6 +21,10 @@ export const atLeastOneMarkedForDeletion = urls => {
     return urls.some(url => url.default.publicationInfo.publicationStatus === 'MARKED_FOR_DELETION' || url.default.mixinTypes.find(mixin => mixin.name === 'jmix:markedForDeletion'));
 };
 
+export const atLeastOneCanonicalLockedForLang = (urls, lang) => {
```
Comment (jsinovassin) [inline core/src/javascript/components/Utils/Utils.jsx:24]: Be careful, The query returning the urls in content editor is not returning the mixinTypes, you could have an null pointer when calling url.default.mixinTypes.find

**Review by jsinovassin — CHANGES_REQUESTED** [review]

File: core/src/javascript/components/AddVanityUrl.jsx (hunk 1 above)
Comment (jahia-carp) [reply core/src/javascript/components/AddVanityUrl.jsx]: Fixed

File: core/src/javascript/components/Utils/Utils.jsx (hunk 2 above)
Comment (jahia-carp) [reply core/src/javascript/components/Utils/Utils.jsx]: Fixed

File: core/src/javascript/components/Utils/Utils.jsx (hunk 3 above)
Comment (jahia-carp) [reply core/src/javascript/components/Utils/Utils.jsx:24]: Fixed

File: core/src/javascript/components/SiteSettingsSeoApp.jsx [hunk 4]
```diff
@@ -324,11 +326,14 @@ class SiteSettingsSeoApp extends React.Component {
     render() {
         let {siteInfo, t, classes, lang, polling: globalPolling} = this.props;
         let polling = !(this.state.publication.open || this.state.move.open || this.state.infoButton.open);
+        let variables = buildTableQueryVariablesAllVanity({selectedLanguageCodes: this.state.loadParams.selectedLanguageCodes, path: siteInfo.path, lang: lang, poll: polling ? SiteSettingsSeoConstants.TABLE_POLLING_INTERVAL : 0, ...this.state.loadParams});
```
Comment (jsinovassin) [inline core/src/javascript/components/SiteSettingsSeoApp.jsx:329]: The poll property is not used in buildTableQueryVariablesAllVanity, it is not necessary to pass it as a param

File: core/src/javascript/components/SiteSettingsSeoApp.jsx [hunk 5]
```diff
@@ -324,11 +326,14 @@ class SiteSettingsSeoApp extends React.Component {
...
 
         return (
             <div className={classes.pageContainer}>
                 <VanityUrlTableData
                     {...this.state.loadParams}
+                    tableQuery={DashboardTableQuery}
+                    variables={variables}
                     path={siteInfo.path}
```
Comment (jsinovassin) [inline core/src/javascript/components/SiteSettingsSeoApp.jsx:337]: Path and lang can be removed here. They are not used in VanityUrlTableData anymore

File: core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx [hunk 6]
```diff
@@ -20,11 +22,14 @@ class SiteSettingsSeoCardApp extends SiteSettingsSeoApp {
...
             this.state.move.open || this.state.infoButton.open);
+        let variables = buildTableQueryVariablesOneNode({selectedLanguageCodes: this.props.languages, path: path, lang: lang, poll: polling ? SiteSettingsSeoConstants.TABLE_POLLING_INTERVAL : 0, ...this.state.loadParams});
 
         return (
             <div>
                 <VanityUrlTableData
                     {...this.state.loadParams}
+                    tableQuery={ContentEditorTableQuery}
+                    variables={variables}
                     path={path}
```
Comment (jsinovassin) [inline core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx:33]: Path and lang can be removed here. They are not used in VanityUrlTableData anymore

File: core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx [hunk 7]
```diff
@@ -20,11 +22,14 @@ class SiteSettingsSeoCardApp extends SiteSettingsSeoApp {
         let {path, t, lang, siteInfo} = this.props;
         let polling = !(this.state.publication.open ||
             this.state.move.open || this.state.infoButton.open);
+        let variables = buildTableQueryVariablesOneNode({selectedLanguageCodes: this.props.languages, path: path, lang: lang, poll: polling ? SiteSettingsSeoConstants.TABLE_POLLING_INTERVAL : 0, ...this.state.loadParams});
```
Comment (jsinovassin) [inline core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx:25]: The polling param is not necessary here

**Review by jsinovassin — CHANGES_REQUESTED** [review]

File: core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx (hunk 7 above)
Comment (jahia-carp) [reply core/src/javascript/components/contentEditor/SiteSettingsSeo/SiteSettingsSeoCardApp.jsx]: Cleaned

File: core/src/javascript/components/SiteSettingsSeoApp.jsx (hunk 5 above)
Comment (jahia-carp) [reply core/src/javascript/components/SiteSettingsSeoApp.jsx]: Cleaned

File: core/src/javascript/components/SiteSettingsSeoApp.jsx (hunk 4 above)
Comment (jahia-carp) [reply core/src/javascript/components/SiteSettingsSeoApp.jsx]: Cleaned thanks

**Review by jsinovassin — DISMISSED** [review]

**Review by jsinovassin — APPROVED** [review]

**Pull request merged**